aerogear/aerogear-metrics-api
# Database Integration

- [x] initial connection and setup logic
- [x] Business logic for inserting records

## Update

This PR ended up turning into a mega PR but this is the initial implementation that allows us to store mobile metrics in the database. There are a number of things to be aware of:


* There is a small amount of validation. The incoming data is parsed into a struct which filters out unnecessary fields. additionally, there are some small checks at the DB layer. ClientId cannot be an empty string and by data field will not accept invalid json. There is no other validation (right now).
* Right now, the client timestamp is being parsed but not being saved. The `event_time` field is populated by Postgres at the time of insert.
* The db_test.go and dao_test.go are actually integration tests. However by default they will not run because of the `// +build integration` build tag added at the top of those files. You can run integration + unit tests using the standard test command, but also passing the flag `tags=integration` @maleck13 Not sure how you feel about this. I found this solution in this [stackoverflow answer](https://stackoverflow.com/questions/25965584/separating-unit-tests-and-integration-tests-in-go). I know it's not as clear as integration tests having their own folder but it's still fairly obvious by looking at the file that the test is an integration test.
* Removed references to timescaledb for now (as suggested by @david-martin )
* Still need to implement some standardised logging approach
* Docs need updating
* Code Coverage at a fairly good 87.6%

The original estimate was 4d and I've spent about 3 days on this so far so there's still time to do some housekeeping around docs, build process, release, etc etc.

## How to verify

fetch this branch and start the database using docker-compose.

Now start the server using `go run cmd/metrics-api/metrics-api.go`

You should see a message like

```
2018/02/22 17:28:30 Starting application... going to listen on :3000
```

Now you can try out the `/metrics` endpoint with the following curl request:

```
curl -X POST \
  http://localhost:3000/metrics \
  -H 'Cache-Control: no-cache' \
  -H 'Content-Type: application/json' \
  -H 'Postman-Token: 76a14738-976f-f0ea-4039-12efeebf4f1e' \
  -d '{
  "clientId": "12345",
  "data": {
  	"app": {
    "id": "com.example.someApp",
    "sdkVersion": "2.4.6",
    "appVersion": "256"
	},
	"device": {
    	"platform": "android",
    	"platformVersion": "27"
	}
  }
}'
```

You should get a the following response:

```
{"timestamp":0,"clientId":"12345","data":{"app":{"id":"com.example.someApp","sdkVersion":"2.4.6","appVersion":"256"},"device":{"platform":"android","platformVersion":"27"}}}
```

Note the `timestamp` field in the response corresponds to the client timestamp which is currently not being handled or saved.

You can log into the database and verify that the record was created:

```
psql -U postgres --host localhost
password: postgres

\c aerogear_mobile_metrics
select * from mobileappmetrics;

           clientid           |          event_time           |                                                                      data
------------------------------+-------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------
 org.aerogear.metrics.testing | 2018-02-22 15:14:03.859053+00 | {"app": {"id": "com.example.someApp", "appVersion": "256", "sdkVersion": "2.4.6"}, "device": {"platform": "android", "platformVersion": "27"}}
```

I'd love to get your reviews @maleck13 @david-martin @wtrocki
